Handle delete events with/out tombstones

tzolov opened the pull request:
- Update Debezium dependencies from 0.9.5.Final to 0.10.0.Final
  - Adjust the EmbeddedEngine class with latest changes in the debezium project.
  - Fix the Consumer<SourceRecord> handler implementation to filter out null SourceRecord. Later are generated on dropTombstones=true.
  - Replace deprecated UnwrapFromEnvelope by ExtractNewRecordState.

  Resolves #6

tzolov committed:
Handle delete events with/out tombstones

  - Update Debezium dependencies from 0.9.5.Final to 0.10.0.Final
  - Adjust the EmbeddedEngine class with latest changes in the debezium project.
  - Fix the Consumer<SourceRecord> handler implementation to filter out null SourceRecord. Later are generated on dropTombstones=true.
  - Replace deprecated UnwrapFromEnvelope by ExtractNewRecordState.

  Resolves #6

artembilan commented on spring-cloud-starter-stream-source-cdc-debezium/pom.xml:
```diff
@@ -95,19 +95,23 @@
 			<groupId>mysql</groupId>
 			<artifactId>mysql-connector-java</artifactId>
 			<version>${version.mysql.driver}</version>
-			<!--<scope>test</scope>-->
```
Why is this not a `test` scope any more?

tzolov replied:
Here i just removed an already commented out scope. It think that it was commented on the first place to replace the mysql-connecter excluded from [debezium-connector-mysql](https://github.com/spring-cloud-stream-app-starters/cdc-debezium/blob/master/spring-cloud-starter-stream-source-cdc-debezium/pom.xml#L45)

artembilan commented on spring-cloud-starter-stream-common-cdc-debezium/spring-cloud-starter-stream-common-cdc-debezium-core/src/main/java/org/springframework/cloud/stream/app/cdc/common/core/CdcStreamConfiguration.java:
```diff
@@ -56,8 +56,17 @@
 
+				if (sourceRecord == null) {
+					logger.debug("Null sourceRecord");
+					return;
```
Why do we just ignore?
Why the same `"null".getBytes()` or `byte[0]` doesn't work for us?

On the other thought: maybe we should emit some `ApplicationEvent` to say that source of events is exhausted and we are not going to receive any update afterwards.

Or do I miss anything?

Thanks

tzolov replied:
Good question which is the core of this issue.  I will try to explain the context and my reasoning the implemented as above. 

Most or perhaps all of the  Debezium connectors have the concept of a `tombstone event` that can be controlled via the `cdc.flattering.dropTombstones` property (drop is true by default). 

The tombstone property controls whether a tombstone event should be generated after a row delete event. When enabled the delete operations are represented by a delete event and a subsequent tombstone event. if the tombstone is disabled only a delete event is sent only.

The tombstone event is particularly important in the context of Kafka. 
> Emitting the tombstone event allows Kafka to completely delete all events pertaining to the given key once the source record got  deleted. 

This raises the question if the tombstone concept is applicable for the EmbeddedDebezium use case we use. But since i don't know the answer i decided not to completely remove this concept although it is disabled by default. 

It turned out though that even if the tombstone is disabled the framework will send a second event to the messageConsumer handler but the value of the `sourceRecord` is null.  This is the reason i've added the ignore section in case of null sourceRecord.  So i had to explicitly filter out the second event. 

Next it turned out that when the tombstone event is enabled the serialization produces a `null` payload and all useful info is encoded into the `cdc.key` header.  But the MessageBuilder doesn't allow null payloads so i've put the "null" string. Which is bad value i guess. 
I should use an empty string instead?

artembilan replied:
Well, thanks, @tzolov , for detailed explanation.

Right, `tombstone` in terms for Apache Kafka really means to remove the `key` from the log.
We use special `KafkaNull` constant and respective `KafkaNullAwarePayloadArgumentResolver` in Spring for Apache Kafka to deal with a `tombstone` concept.

Not sure if we need to introduce something similar here to have as much logic coverage as possible, but this some kind of food for you to think more.

An empty string is going to be treated as a `null payload` in some method argument resolver and therefore a logic might not be called again. But at the same time the `byte[0]` is treated same way:

```
	/**
	 * Specify if the given {@code payload} is empty.
	 * @param payload the payload to check (can be {@code null})
	 */
	protected boolean isEmptyPayload(@Nullable Object payload) {
		if (payload == null) {
			return true;
		}
		else if (payload instanceof byte[]) {
			return ((byte[]) payload).length == 0;
		}
		else if (payload instanceof String) {
			return !StringUtils.hasText((String) payload);
		}
		else {
			return false;
		}
	}
```

That's why I suggest to emit an `ApplicationEvent` instead of attemting to send a message with an *empty payload*.

Thanks

artembilan reviewed (changes requested):
I have left some review.

Bare with me that I'm not familiar with CDC 😄

tzolov committed:
Imporve the Tombstones handling logic. add tests

artembilan commented on spring-cloud-starter-stream-source-cdc-debezium/src/test/java/org/springframework/cloud/stream/app/cdc/debezium/source/CdcSourceIntegrationTests.java:
```diff
@@ -92,6 +98,113 @@ public void testOne() throws InterruptedException {
+				received = messageCollector.forChannel(this.channels.output()).poll(10, TimeUnit.SECONDS);
+				if (received != null) {
+					System.out.println("Headers: " + received.getHeaders());
+					System.out.println("Payload: " + received.getPayload());
```
I think you wanted to *assert* messages here instead of printing them into SOUT...
Or just remove this `do...while` at all...

tzolov replied:
Good point. I will change the approach to just drain the queue and assert the size.

artembilan commented on spring-cloud-starter-stream-source-cdc-debezium/src/test/java/org/springframework/cloud/stream/app/cdc/debezium/source/CdcSourceIntegrationTests.java:
```diff
@@ -92,6 +98,113 @@ public void testOne() throws InterruptedException {
+				received = messageCollector.forChannel(this.channels.output()).poll(10, TimeUnit.SECONDS);
+				if (received != null) {
+					System.out.println("Headers: " + received.getHeaders());
+					System.out.println("Payload: " + received.getPayload());
```
DITTO

tzolov committed:
Address review remark to get rid of the do while loops

tzolov committed:
Add logic to sed KafkaNull only if the spring-kafka is present on the classpath and DBZ tombstones are enabled. If kafka is not present the tombstone messages are ignored.  Add a separate project to test the behavior whit kakfa on the classpath.

tzolov committed:
Address open review comments

 - Set the messageKey to headers['cdc_key'] by default.
 - Add support for the operation.header and add.source.fields flattering properties.
 - Add CdcFlatteringIntegrationTest - for testing the flattering and none flattering response representations for different message types.
 - Refactor all integration tests to use ApplicationContextRunner. Later allows to exclude Kafka from the classpath for testing special tombsotne behavior.
 - Replace dot by underscore in header kyes. E.g. cdc_key, cdc_topic, cdc_operation and cdc_offset
 - CdcTombstoneConfiguration addlow to (re)send Tombstone events using KafkaNull payload. Activated only if Kafka is present on the classpath. If not the tombstone events are ignored.

artembilan commented on spring-cloud-starter-stream-common-cdc-debezium/spring-cloud-starter-stream-common-cdc-debezium-core/src/main/java/org/springframework/cloud/stream/app/cdc/common/core/CdcStreamConfiguration.java:
```diff
@@ -92,4 +121,24 @@
+	@Bean
+	@ConditionalOnMissingBean(CdcTombstoneConfiguration.TombstoneSupport.class)
+	public CdcTombstoneConfiguration.TombstoneSupport nullTombstoneHelper() {
+		return new CdcTombstoneConfiguration.TombstoneSupport() {
```
That's something too complicated. 
I would load a `KafkaNull` instance into some `static` property if it is present and would initialize it with `null` otherwise.
So, your code above would be just like this:
```
if (cdcJsonPayload == null) {
	cdcJsonPayload = KAFKA_NULL;
}
```

tzolov replied:
Actually if you remember i did implement such code (using reflection).  But static code using reflection doesn't play well for testing.  
Hit the wall while writing tests with the `ApplicationContextRunner().withClassLoader(new FilteredClassLoader(KafkaNull.class))`.  While the test context created like this will exclude the KafkaNull.class, the static implementation would still load it and show it as present on the classpath.

artembilan commented on spring-cloud-starter-stream-common-cdc-debezium/spring-cloud-starter-stream-common-cdc-debezium-core/src/main/java/org/springframework/cloud/stream/app/cdc/common/core/CdcTombstoneConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+ * @author Christian Tzolov
+ */
+@Configuration
+public class CdcTombstoneConfiguration {
```
As I said in the other comment, it doesn't look like we need this configuration at all...

tzolov replied:
check my comment above ;)

artembilan commented on spring-cloud-starter-stream-common-cdc-debezium/spring-cloud-starter-stream-common-cdc-debezium-core/src/main/java/org/springframework/cloud/stream/app/cdc/common/core/CdcTombstoneConfiguration.java:
```diff
@@ -0,0 +1,90 @@
+	@StreamMessageConverter
+	@ConditionalOnClass(name = ORG_SPRINGFRAMEWORK_KAFKA_SUPPORT_KAFKA_NULL)
+	public MessageConverter kafkaNullConverter() {
+		class KafkaNullConverter extends AbstractMessageConverter {
```
I wonder why do we need this extension.
Spring Kafka is able to handle `KafkaNull` very well.
Also it doesn't look like you handle it at all.
In terms of Spring Kafka a `KafkaNull` turns to *real* `null` during sending a record.
So, what is the point of this one, please?

tzolov replied:
Actually it doesn't. Without this i am getting  conversion exceptions: 
```
Caused by: java.lang.IllegalStateException: Failed to convert message: 'GenericMessage [payload=org.springframework.kafka.support.KafkaNull@141d15c, headers={cdc_topic=my-app-connector.inventory.customers, id=c8bdf0eb-600b-efc4-4a71-32a67190f2c4, contentType=application/json, cdc_key={"id":1010}, timestamp=1576682874718}]' to outbound message.
	at org.springframework.cloud.stream.binding.MessageConverterConfigurer$OutboundContentTypeConvertingInterceptor.doPreSend(MessageConverterConfigurer.java:389)
```

artembilan reviewed (changes requested):
Some review for the current code.

Thanks

tzolov committed:
simplify kafkanull handling

artembilan commented on cdc-debezium-app-dependencies/pom.xml:
```diff
@@ -13,17 +13,19 @@
-		<!-- Note: postgresql version MUST match the version used by
-		the debezium postgres connector -->
+<!--		<spring-cloud-stream.version>2.1.2.RELEASE</spring-cloud-stream.version>-->
+		<version.debezium>0.10.0.Final</version.debezium>
```
Any chances that we can upgrade already: https://debezium.io/blog/2019/12/18/debezium-1-0-0-final-released/ ?

tzolov replied:
No

tzolov replied:
it took 1 year to get to this point. The upgrade will come with next version :) probably after an year :)

tzolov committed:
Update to debezium 1.0
